functorch/csrc/BatchRulesRandomness.cpp

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <ATen/ATen.h>
+#include <functorch/csrc/DynamicLayer.h>
+#include <functorch/csrc/BatchRulesHelper.h>
+
+namespace at {
+namespace functorch {
+
+void check_randomness(std::string randomness) {
+  TORCH_CHECK(
+    randomness != "error",
+    "vmap: called random operation while in randomness error mode. Please either use the "
+    "'same' or 'different' randomness flags on vmap or perform the randomness operation out of vmap"
+  );
+}
+
+template <typename F, F Func, typename... ExtraArgs>
+Tensor random_batching_rule(IntArrayRef shape, ExtraArgs... extra_args) {
+  c10::impl::ExcludeDispatchKeyGuard guard(kVmapModeKey);
+  auto maybe_layer = maybeCurrentDynamicLayer();
+  VmapDimVector shapeVec(shape.begin(), shape.end());
+  shapeVec.insert(shapeVec.begin(), maybe_layer->batchSize());
```

Comment:
insertion at the beginning is O(n), should start with [batch_size] and then insert shape.